Add preliminary doc on calibration metrics

dstrain115 opened the pull request:
- Explains different metrics and how to use them

mpharrigan commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,170 @@
+  print(metric_name)
+  print('------')
+  for qubit_or_pair in latest_calibration[metric_name]:
+     # Note that although the value is often singular,
```
p.s. this is annoying

dstrain115 replied:
Acknowledged.

mpharrigan commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,170 @@
+[thesis](https://web.physics.ucsb.edu/~martinisgroup/theses/Chen2018.pdf).
+
+ 
+# T1 
```
I find your heading-level strategy mildly confusing. Consider having all these headings be at the same level?

dstrain115 replied:
Yeah, I have no idea what I was doing here.  Made all the metrics '###' heading.

mpharrigan commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,170 @@
+# Replace YOUR_PROJECT_ID with the id from your cloud project.
+engine = cg.Engine(project_id=YOUR_PROJECT_ID, proto_version=cg.ProtoVersion.V2)
+
+# TODO(#2720): this PR refactors some of these functions...
```
do you want to leave this in?

dstrain115 replied:
Removed and coordinated with @lindmarkm to update this once this PR goes in.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,170 @@
+calibration_name = ("projects/" + PROJECT_ID 
+                    + "/processors/" + PROCESSOR_ID
+                    + "/calibrations/" 
+                    + str(EPOCH_TIMESTAMP_SECONDS)) 
```
It's a bit unfortunate that we expose REST protocol to the users like that.

dstrain115 replied:
I think PR 2720 should help alleviate some of this.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,170 @@
+*   Metric key: single_qubit_readout_p0_error
+
+The p00 readout error is the probability that a qubit is incorrectly measured as
+|1> after being prepared in the |0> state.  This is mostly a measure of the
```
Use Unicode: |0⟩. Applies throughout.

dstrain115 replied:
Done

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,170 @@
+
+Single qubit gate error is estimated using randomized benchmarking by taking
+sequences of varying length of the 24 gates within the Clifford group
+(those generated by the Pauli gates {±X,±Y,±Z}) then applying the inverse of the
```
Clifford group is not generated by the Pauli gates. You can instead write "(those that preserve the Pauli group under conjugation)".

dstrain115 replied:
Changed

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,170 @@
+produce the total error.  This error is calculated for one qubit at a time while
+all other qubits on the device are idle (isolated).
+
+The purity error is a measure of the incoherent error (as opposed to
```
I think people generally just speak of "purity" when they mean the trace of the square of the density matrix. Is this the quantity we report? If so, we should drop "error". If not, we should explain what it is we actually report.

dstrain115 replied:
This metric is 1-tr(p^2), so one minus purity.  Elaborated this in the docs.

viathor replied:
Assuming that by p you mean the density matrix, this is then the (unnormalized) [linear entropy](https://en.wikipedia.org/wiki/Linear_entropy). It'd be good to use the established terminology.

Note: I'd indicate explicitly whether it is normalized or not since both are used. Normalized linear entropy is given by the formula (1 - tr(p^2)) * 2^n / (2^n - 1) and has values in [0, 1]. Unnormalized linear entropy is given by the formula 1 - tr(p^2) and has values in [0, 1 - 2^-n].

dstrain115 replied:
Ok, I am not actually sure.  I will go find out before I write something incorrect.

viathor replied:
Thanks! Please ping this PR when it's ready for another look.

dstrain115 replied:
I added a separate section on total vs purity error so I could elaborate further.  Can you review this and make sure I didn't say anything incorrect?

dstrain115 replied:
Adjusted this section to include average vs pauli vs purity error.  Can you reevaluate to make sure I didn't say anything stupid?

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,168 @@
+*   Metric key: single_qubit_rb_total_error
+*   Metric key: single_qubit_rb_purity_error
+
+Single qubit gate error is estimated using randomized benchmarking by taking
```
Unfortunately, there is more than one error measure in the literature. To avoid confusion, we should state explicitly whether we're reporting average error, Pauli error, depolarization error?

dstrain115 replied:
It's average error.  Moved this to its own section to add more detail.

sboixo replied:
A few months ago we were using Pauli error instead of "average error" (1- average fidelity) because that's the right quantity for the discrete error model (average error needs to renormalized depending on the circuit size). 
Please considering reporting Pauli errors instead of average error.

dstrain115 replied:
I have adjusted everything to use average error instead total error.  I have also Pauli error to the metrics.

Note that the documentation is now slightly forward looking, since these metrics will take a week or two to get into the engine code and to be released.

kevinsung commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,186 @@
+
+Note that the metric names below are subject to change without notice.
+
+### P_00 readout error
```
Why two 0s, when the metric key only has one?

dstrain115 replied:
Fixed.  Note that both will appear for a while until I get everyone to switch over.

kevinsung replied:
But I think having just one 0 makes sense. I don't know what the second 0 means. P_00 could be interpreted as the probability of measuring a 0 given that you prepared a 0. But that's not what it means: it is the probability of measuring a 1!

dstrain115 replied:
The idea is that the experiment is to prepare the state in 0, then measure it in 0, hence 00.  The error is a combination of both state preparation and measurement, so it's 1-p_00.  Though, now that I think about it, it does not include the rare case of the state being prepared incorrectly combined with an incorrect measurement.

People around here are coalescing around calling this p00 and p11 error, but I can raise it in the working group to see if there is a preference between p0 and p1 as well.

kevinsung replied:
I see, that makes sense.

kevinsung commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,199 @@
+### P_11 readout error
+*   Metric key: single_qubit_readout_p11_error
+
+The P1 readout error is the probability that a qubit is incorrectly measured as
```
Again, you have P1 where every where else it is P11. But also again, I think P1 makes more sense, because I don't know what the second 1 means. Perhaps you can explain?

95-martin-orion commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+ 
+*   Metric key: two_qubit_sqrt_iswap_gate_xeb_cycle_average_error
+*   Metric key: two_qubit_sqrt_iswap_gate_xeb_cycle_pauli_error
+*   Metric key: two_qubit_sycamore_gate_xeb_cycle_purity_error
```
Duplicate `two_qubit_sycamore_gate_xeb_cycle_purity_error` key.
(`_sycamore_` -> `_sqrt_iswap_`?)

dstrain115 replied:
Done.

95-martin-orion commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+### 2-qubit Isolated XEB error
+ 
+*   Metric key: two_qubit_sqrt_iswap_gate_xeb_cycle_average_error
+*   Metric key: two_qubit_sqrt_iswap_gate_xeb_cycle_pauli_error
```
Which of these is the correct meaning for two-qubit Pauli error values?
- Given a qubit involved in `<two-qubit gate>`, `<value>` is the chance that the qubit experiences a Pauli error.
- Given an instance of `<two-qubit gate>`, `<value>` is the chance that _at least_ one qubit involved in the gate experiences a Pauli error.

dstrain115 commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+then query for the current or previous state of the calibration.
+Calibrations are also available for past jobs.
+
+Calibration objects are a dictionary from metric name (see below) to the value
```
Done.

dstrain115 commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+engine = cg.Engine(project_id=YOUR_PROJECT_ID, proto_version=cg.ProtoVersion.V2)
+processor = engine.get_processor(processor_id=PROCESSOR_ID)
+
+# gets the latest calibration metrics
```
Done.

dstrain115 commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+# If you would like to find a calibration from a time-frame:
+calibration_list = processor.list_calibration(START_SECONDS, END_SECONDS)
+
+# If you know the job-id, you can retrieve the calibration that the job used.
```
Done.

dstrain115 commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+
+Several metrics below define average error, pauli error and purity error.
+The average error measures the error ε_a, averaged over all possible input
+states.  
```
Changed to use your suggestion.

dstrain115 commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+
+The [Purity](https://en.wikipedia.org/wiki/Purity_(quantum_mechanics)) of the
+quantum state is often defined as the trace of the density matrix of the
+state after the operations (RB or XEB) have been applied.  The purity error,
```
Done.

dstrain115 commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+or impurity, is one minus the purity, scaled to be from zero to one.  This is
+often referred to as the normalized
+[Linear Entropy](https://en.wikipedia.org/wiki/Linear_entropy).
+after the randomized benchmark circuit has been applied.  
```
Removed.

dstrain115 commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+result.  
+
+There are several sources of error in this model.  This error is primarily
+composed of error in measurement (read out) of the qubit while in the ground
```
Done.

dstrain115 commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+being prepared in the |1⟩ state.  The p_11 error is defined as one minus this
+result.  
+
+This is dominated by the error in measurement (read out) of the qubit, but it
```
Done.

dstrain115 commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+Note that the metric names below are subject to change without notice.
+
+### P_00 readout error
+
```
Done.

dstrain115 commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+*   Metric key: single_qubit_readout_separation_error
+
+When measured by the system, the |0⟩ and |1⟩ states manifest as outgoing analog
+signals.  These signals must be interpreted as being in one state or the other.
```
Done.

dstrain115 commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+Single qubit gate error is estimated using randomized benchmarking by taking
+sequences of varying length of the 24 gates within the Clifford group
+(those that preserve the Pauli group under conjugation) then applying the
+inverse of the expected final state.  The result of this sequence should always
```
Changed, but I kept the next sentence since maybe it is not obvious to everyone.

dstrain115 commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+*   Metric key: two_qubit_sycamore_gate_xeb_cycle_purity_error
+
+Two qubit error is primarily characterized by cross-entropy benchmarking (XEB).
+This is computed by performing a "cycle" of a random one-qubit gate on each
```
Done.

dstrain115 commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+this experiment on their own if they have a specific layout commonly used in
+their experiment.
+
+
```
Done.

dstrain115 commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+and depolarization error.
+
+The [Purity](https://en.wikipedia.org/wiki/Purity_(quantum_mechanics)) of the
+quantum state is often defined as the trace of the density matrix of the
```
Done.

zchen088 commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,200 @@
+document for a description and comparison between average error, Pauli error,
+and depolarization error.
+
+The [Purity](https://en.wikipedia.org/wiki/Purity_(quantum_mechanics)) of the
```
I would consider referring instead to the "unitarity", which is the decay rate per gate (or cycle) of the purity. We choose to normalize this decay rate to be on the same scale as the "total error" (which is also a rate per gate). You can mention that in casual parlance we often refer to the unitarity as the "purity error".

dstrain115 replied:
Done.

kevinsung commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,199 @@
+*   Metric key: two_qubit_sycamore_gate_xeb_cycle_incoherent_error
+
+Two qubit error is primarily characterized by applying cross-entropy
+benchmarking (XEB).  This procedure consists of performing a "cycle" of a
```
Worth emphasizing that the error is of the cycle, and not the two-qubit gate (right??)

dstrain115 replied:
Done.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+related by a multiplicative factor based on the number of qubits.
+
+See Table 1 on page 11 of the
+[https://arxiv.org/abs/1910.11333](Supplementry Information)
```
Link title and url need to be swapped.

dstrain115 replied:
Oops!  Fixed.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+# Calibration Metrics
+
+Processors that are quantum computers periodically undergo calibrations to
```
Simplify by replacing "Processors that are quantum computers" with "Quantum processors".

dstrain115 replied:
Done.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+# Calibration Metrics
+
+Processors that are quantum computers periodically undergo calibrations to
+maintain the quality of the programs that can be run on these processors.
```
nit: use pronouns to avoid repetition: s/on these processors/on them/

dstrain115 replied:
Done.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+  print('------')
+  for qubit_or_pair in latest_calibration[metric_name]:
+     # Note that although the value is often singular,
+     # the metric_value is of the type list and can have multiple values     
```
nit: need dot at EOL

dstrain115 replied:
Done.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+
+## Average, Pauli and Incoherent Error
+
+Several metrics below define average error, pauli error and incoherent error.
```
nit: capitalize Pauli

dstrain115 replied:
Done.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+## Average, Pauli and Incoherent Error
+
+Several metrics below define average error, pauli error and incoherent error.
+This section explains the difference between each of those metrics.
```
nit: s/those/these/ (since you just mentioned them)

dstrain115 replied:
Done

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+X, Y, or Z.  If the errors are distributed in a uniform distribution across all
+three axes, the probability of applying an erroneous Pauli gate X, Y, or Z will
+be the Pauli error divided by three.  The Pauli error and average error are
+related by a multiplicative factor based on the number of qubits.
```
nit: s/based on/dependent on/

dstrain115 replied:
Done.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+input states.
+
+Pauli error defines decoherence of a single qubit in one of the Pauli channels
+X, Y, or Z.  If the errors are distributed in a uniform distribution across all
```
nit: s/a uniform/ the uniform/ (there exists only one uniform distribution over three elements)

dstrain115 replied:
Done.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+input states.
+
+Pauli error defines decoherence of a single qubit in one of the Pauli channels
+X, Y, or Z.  If the errors are distributed in a uniform distribution across all
```
nit: s/across/over/

dstrain115 replied:
Done.

viathor commented on docs/google/calibration.md:
````diff
@@ -0,0 +1,217 @@
+Each metric can be referenced by its key in the calibration object, e.g.
+```latest_calibration['single_qubit_idle_t1_micros']```.
+
+**Note that the metric names below are subject to change without notice.**
````
Do we plan to freeze them at some point, e.g. cirq 1.0? Would it make sense to promise this here?

dstrain115 replied:
I don't think there is a plan to freeze this, and they will likely change drastically as devices change and evolve.  For instance, if the number of qubits increases, we may no longer be able to provide isolated metrics.  I don't think we want to make any promises beyond the promise that they will almost definitely change.

viathor replied:
As usual, this is unlikely to stop users from writing code that depends on this (e.g. predictive models for circuit fidelity). That said, I'm glad we're reserving the right to change this.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+impossible to distinguish.  This is classified as the separation error, and is
+calculated by fitting Gaussian distributions to the signals prepared in the
+|0⟩ state and |1⟩ state and calculating the overlap between the two
+distributions.
```
I still think that our clients won't know how to use this. What's the goal in exposing this metric?

dstrain115 replied:
This could potentially be discussed at the SPWG core meeting.  For now, we still expose this metric, so I'll leave the explanation in.  We can always delete it if we remove the metric.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+
+Single qubit gate error is estimated using randomized benchmarking by taking
+sequences of varying length of the 24 gates within the Clifford group
+(those that preserve the Pauli group under conjugation) then applying then
```
nit: repeated words: "then applying then applying the inverse", should be "then applying the inverse"

dstrain115 replied:
Done.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+(those that preserve the Pauli group under conjugation) then applying then
+applying the inverse of the unitary equivalent to the executed gate sequence.
+The result of the total sequence should always be the identity (|0⟩ state).
+The final error is measured and compared against this
```
nit: underflow on this line

dstrain115 replied:
Done.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+*   Metric key: single_qubit_idle_t1_micros
+
+The T1 of a qubit represents the time constant of the exponential decay of a
+qubit in the excited |1⟩ state to naturally decay into the ground |0⟩ state.
```
nit: Remove "naturally decay into". The resulting sentence is clearer, doesn't repeat the word "decay", avoids the vague term "naturally" and reads better overall.

dstrain115 replied:
Done.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+
+The T1 of a qubit represents the time constant of the exponential decay of a
+qubit in the excited |1⟩ state to naturally decay into the ground |0⟩ state.
+This is calculated by preparing an excited state with a microwave pulse
```
nit: "an excited state" -> "the excited state" since we do mean a specific one (namely, the first one). If you'd like extra clarify you can write "by preparing a qubit in the excited state".

dstrain115 replied:
Done.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+*   Metric key: two_qubit_sycamore_gate_xeb_cycle_incoherent_error_per_cycle
+
+Two qubit error is primarily characterized by applying cross-entropy
+benchmarking (XEB).  This procedure consists of performing a "cycle" of a
```
nit: s/consists of/consists in/

dstrain115 replied:
Consists in sounds weird to me.  I just got rid of it and changed to "This procedure performs..."

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+*   Metric key: two_qubit_sycamore_gate_xeb_cycle_incoherent_error_per_cycle
+
+Two qubit error is primarily characterized by applying cross-entropy
+benchmarking (XEB).  This procedure consists of performing a "cycle" of a
```
"performing" -> "repeatedly performing" to avoid the implication that only one cycle is performed.

dstrain115 replied:
Done.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+random one-qubit gate on each qubit followed by the two qubit entangling gate.
+The resulting distribution is analyzed and compared to the expected distribution
+using cross entropy.  The value reported is the error rate per cycle (both
+the 1 qubit gates as well as the 2 qubit gate) is incorrect.
```
The "is incorrect" seems out of place. Remove?

dstrain115 replied:
Done.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+using cross entropy.  The value reported is the error rate per cycle (both
+the 1 qubit gates as well as the 2 qubit gate) is incorrect.
+
+See the above section for descriptions of average, puali, and incoherent error.
```
nit: "puali" -> "Pauli" (swap a, u and capitalize p)

dstrain115 replied:
Done.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+    two_qubit_parallel_sycamore_gate_xeb_cycle_incoherent_error_per_cycle
+
+These metrics are calculated the same way as the 2-qubit isolated XEB error
+metrics.  However, this metric calculates the error of multiple 2-qubit cycles
```
"multiple" -> "multiple parallel" for clarity.

dstrain115 replied:
Done.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+    two_qubit_parallel_sycamore_gate_xeb_cycle_incoherent_error_per_cycle
+
+These metrics are calculated the same way as the 2-qubit isolated XEB error
+metrics.  However, this metric calculates the error of multiple 2-qubit cycles
```
nit: "metric calculates" -> "metric quantifies"

dstrain115 replied:
Much better verb.  Done.

viathor commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+
+These metrics are calculated the same way as the 2-qubit isolated XEB error
+metrics.  However, this metric calculates the error of multiple 2-qubit cycles
+at a time.  Four different discrete patterns of 2-qubits are calculated, with
```
nit: "are calculated" -> "are used"

dstrain115 replied:
Done.

peterse commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,217 @@
+qubit is measured at a time), while the parallel error is taken for all qubits
+at the same time.
+
+### Readout separation error
```
Is this value accounted for in p_00, p_11? The way I'm interpreting the relationship is
```
p_00 = P("measured as |0⟩" | "prepared in the |0⟩ state")
         = P("measured as |0⟩" | "appeared in |0> region of IQ plane")P("appeared in |0> region of IQ plane" |"prepared in the |0⟩ state")
          = F_s P("appeared in |0> region of IQ plane" |"prepared in the |0⟩ state")
```
Otherwise it seems like I'd need to propagate error proportional to (1 - F_s) for any analysis that consumes p_00/p_11

dstrain115 replied:
Yes, for p_00 and p_11, we cannot tell why the error occurred (in and of itself), whether it was decay, measurement error, or statistical error.

This metric purely measures the error we would expect if both |1> and |0> are Gaussian distributions sampled without other error.  (Since Gaussians are infinite, there will always be some, likely small, overlap)

Added a sentence to clarify this.

dstrain115 commented on docs/google/calibration.md:
```diff
@@ -0,0 +1,204 @@
+  print('------')
+  for qubit_or_pair in latest_calibration[metric_name]:
+     # Note that although the value is often singular,
+     # The metric_value is of the type list and can have multiple values     
```
Done

dstrain115 commented:
> I just noticed that the remaining unaddressed comments are exactly the comments hidden by github, so sending a heads-up that there is 12 comments left :-)

Oops.  I missed those.  There are too many comments, I am losing track!  I think I addressed them all now.
